warehouse/forklift/decorators.py

```diff
@@ -0,0 +1,127 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import cgi
+
+from pyramid.httpexceptions import HTTPBadRequest, HTTPForbidden
+
+from warehouse.admin.flags import AdminFlagValue
+from warehouse.forklift.utils import _exc_with_message
+
+
+def sanitize(wrapped):
+    """
+    Wraps a Pyramid view to sanitize the incoming request of "bad" values.
+
+    There's a lot of garbage that gets sent to the upload view, which we'll
+    sanitize to clean that up and/or fail early rather than getting failures
+    deeper in the stack.
+    """
+
+    def wrapper(context, request):
+        # Do some cleanup of the various form fields, there's a lot of garbage that
+        # gets sent to this view, and this helps prevent issues later on.
+        for key in list(request.POST):
+            value = request.POST.get(key)
+            if isinstance(value, str):
+                # distutils "helpfully" substitutes unknown, but "required" values
+                # with the string "UNKNOWN". This is basically never what anyone
+                # actually wants so we'll just go ahead and delete anything whose
+                # value is UNKNOWN.
+                if value.strip() == "UNKNOWN":
+                    del request.POST[key]
+
+                # Escape NUL characters, which psycopg doesn't like
+                if "\x00" in value:
+                    request.POST[key] = value.replace("\x00", "\\x00")
+
+        # Check if any fields were supplied as a tuple and have become a
+        # FieldStorage. The 'content' field _should_ be a FieldStorage, however,
+        # and we don't care about the legacy gpg_signature field.
+        # ref: https://github.com/pypi/warehouse/issues/2185
+        # ref: https://github.com/pypi/warehouse/issues/2491
+        for field in set(request.POST) - {"content", "gpg_signature"}:
+            values = request.POST.getall(field)
+            if any(isinstance(value, cgi.FieldStorage) for value in values):
```

Comment:
Thinking out loud - `cgi` is deprecated and going to be removed in 3.13 - are we still getting things coming in as FieldStorage instances? Might be good to insert a sentry message to try and track down if this is still happening. Not a merge blocker, but would be happy to re-review something like that.